dapps/profile.vue/src/components/did/DidService.ts

```diff
@@ -1,93 +1,124 @@
-       this.didDocument = await didInstance.getDidDocument(did);
-       // this.didDocument = await this.runtime.did.getDidDocument(did); */
+  async fetchDidDocument(): Promise<DidDocument> {
+    const did = await this.runtime.did.convertIdentityToDid(this.runtime.activeIdentity);
 
-    // console.log('didDocument', this.didDocument);
+    return this.runtime.did.getDidDocument(did);
+  }
 
-    return Promise.resolve(TEST_DID_DOC);
+  async updateDidDocument(newDidDoc: DidDocument): Promise<void> {
+    return this.runtime.did.setDidDocument(newDidDoc.id, newDidDoc);
   }
 
-  static async getDelegates(didDocument: DidDocumentTemplate): Promise<Delegate[]> {
+  /**
+   * Returns the delegates excluding the owner entry
+   * @param didDocument specified did document
+   */
+  async getDelegates(didDocument: DidDocument): Promise<Delegate[]> {
     // regex to remove #key suffixes
     const regex = /([#])(key-)(\d)+/;
 
     return Promise.all(didDocument.authentication
-      .map((item) => item.replace(regex, ''))
-      .filter((item) => item !== didDocument.id)
+      // Filter out owner
+      .filter((auth) => auth !== DidService.getOwner(didDocument))
+      .map((auth) => auth.replace(regex, ''))
       .map(async (delegateId) => ({
         did: delegateId,
-        /* TODO getUserAlias doesn't handle DIDs yet
-           note: await profileUtils.getUserAlias(this.getRuntime(), delegateId), */
-        note: await Promise.resolve('MOCK NOTE'),
+        note: await profileUtils.getUserAlias(this.runtime, delegateId),
       })));
   }
 
-  static saveDelegates(delegates: Delegate[]): void {
-    // TODO logic
-    console.log('Updating delegates...', delegates);
+  static updateDelegates(didDoc: DidDocument, delegates: Delegate[]): DidDocument {
+    // Preserve the owner
+    const ownerEntry = DidService.getOwner(didDoc);
+    const newDoc = didDoc;
+    newDoc.authentication = [
+      ownerEntry,
+      ...delegates.map((delegate) => delegate.did),
+    ];
+    return newDoc;
   }
 
-  static async getServiceEndpoints(didDocument: DidDocumentTemplate): Promise<ServiceEndpoint[]> {
+  // eslint-disable-next-line class-methods-use-this
+  getServiceEndpoints(didDocument: DidDocument): ServiceEndpoint[] {
+    if (!didDocument.service) {
+      return [];
+    }
     return didDocument.service.map((endpoint) => ({
       label: endpoint.id,
       url: endpoint.serviceEndpoint,
     }));
   }
 
-  static saveServiceEndpoints(endpoints: ServiceEndpoint[]): void {
-    // TODO logic
-    console.log('Updating endpoints...', endpoints);
+  static updateServiceEndpoints(didDoc: DidDocument, endpoints: ServiceEndpoint[]): DidDocument {
+    return {
+      ...didDoc,
+      service: endpoints.map((endpoint) => ({
+        id: endpoint.label,
+        serviceEndpoint: endpoint.url,
+        type: 'TODO', // TODO
+      })),
+    };
+  }
+
+  /**
+   * TODO: Currently we don't implement controller logic. Return empty array
```

Comment:
Why we don't implement controller logic? ;)

Reply:
We agreed on it during the planning. Currently we don't plan to have multiple controllers I suppose, but I don't really remember. See the note in figma :)